### ISSUE TITLE
Possible issue with MRS/MSR instructions...

### ISSUE THREAD

**Issue by ElectronAsh** [krevanth/ZAP#132]
Hi,

Firstly, thank you for writing this core.

I've been using it to try to boot the 3DO BIOS, for a future FPGA core.
Out of the five or so open-source ARM cores I tried, Zap is the only one which can boot the BIOS far enough to display the 3DO logo.

I think there might be a possible issue with the MRS/MSR instructions, though.
I could be entirely wrong. I'm not super experienced with debugging this. lol

I'll try to explain the issue as best I can...

I'm currently using the version of Zap just before the commit for the SystemVerilog version.
When I tried the latest SV version, it would crash after only a handful of instructions.
Verilator does support SV, but it might have some issues there, or I'm doing something else wrong with the timings or C code setup.

I'm running it as a sim, using Verilator along with the Dear Imgui library.
The whole of Zap is in Verilog, whereas higher-level things like RAM and BIOS access are handled in C.

In this screenshot, it's showing an MRS(AL) instruction which was just loaded into the decode stage (lower-left).

I added some code which shows the source register in Green, and the destination register in Red.
That looks correct for reading from SVC_SPSR (green), and writing back to R8 (red)...

(the CPU is in Supervisor mode at this point. You can see the lower bits of the CPSR register in the left-hand window are b10011, so the MRS instruction should be grabbing the SPSR. The Green and Red show those source and dest regs are correct during the decode stage.)

![sim_3do msr instruction](https://user-images.githubusercontent.com/24658385/205460121-c130ed35-3002-45e5-b5ce-a2562a19d79e.png)

That seems to tally fine with the regfile index numbers in zap_localparams.vh...

![zap_localparams_vh](https://user-images.githubusercontent.com/24658385/205460258-e0fa27d4-89b2-4954-97df-7c4a1044bdbd.png)

But, the wrong value (0x60000013) gets written back to the R8 reg once the MRS instruction reaches the rb (Writeback?) stage of the pipeline.

My current theory is that the "translate" function in zap_functions.vh isn't being directly "called" by the decode_mrs task in zap_decode.v, so the index value in ARCH_CURR_SPSR isn't being updated correctly once the MRS instruction reaches the Writeback stage.

So instead of writing the contents of the PHY_SVC_SPSR register to R8, it's grabbing PHY_FIQ_SPSR or something else at the time the Writeback happens.

As I say, I could be completely wrong here. It might even be an issue with how I have the clock and Verilator set up.
Or some edge case bug in Verilator itself, which I have seen in the past.

(I noticed that for the actual regfile access elsewhere in Zap, it does call the "translate" function, so ARCH_CURR_SPSR gets updated accordingly. The "decode_data_processing" task also uses the ARCH_CURR_SPSR index value directly)

I know you have some regression test code which I might try to get running later.
Does this theory of a possible bug seem plausible, or does MRS and MSR work OK for you in all CPU modes?

I should also mention that the 3DO uses an ARM60 CPU, and the pin for Big Endian mode is tied High on the 3DO motherboard.

I know Zap doesn't officially support Big Endian mode yet, so I had to do some tweaks to get it running this far.
(with help from "fixel", to figure out which parts of the Verilog to tweak.)...
https://github.com/ElectronAsh/sim_3do/blob/2aa727aef7f53444558ef03000538b47a8f07f84/rtl/zap/zap_memory_main.v#L237

I realize this could be creating other issues with the core, along with the fact that the ARM60 is ARMv4, but Zap is ARMv5TE.

Thanks in advance,
ElectronAsh.

**Comment by ElectronAsh** [comment]
Not sure if this helps with the possible "bug" report, but I'm seeing quite a few of these messages just before the crash...

![3DO BIOS SWI Overrun crash, Zap core messages](https://user-images.githubusercontent.com/24658385/205461151-586c123a-641c-4e89-b383-2816e6379a2b.png)

So this could be related to the fact that a FIQ is triggered by the 3DO logic just as the MRS instruction happens.
As far as I can tell, the CPU is in Supervisor mode when the MRS instruction goes through the whole pipeline, so I think that part is OK.

The 3DO BIOS is also clearing the pending FIQ bits during its FIQ handler, so that looks OK.

I'm not sure exactly what the other messages from Zap mean, though?
Are they just a warning that the pipeline is getting cleared when a write to R15/PC is seen?

(I know the ARMv5 will switch to Thumb mode as well, if the LSB bit of the PC is set during a direct write to the PC, or something. That doesn't seem to be happening here, which is good.)

ElectronAsh.

**Comment by krevanth** [comment]
Thanks @ElectronAsh . I will take a look.

**Comment by ElectronAsh** [comment]
I managed to get the very latest SV version of Zap starting to run further.

No 3DO logo just yet, as it's getting stuck in a routine near the start of the BIOS.

The main problem there was that I was basically asserting i_wb_ack the very instant I saw o_wb_stb go high (in the C code), which obviously isn't something the logic would normally see.

Now it will properly wait for one clock cycle before asserting ack after seeing stb.
It doesn't seem to slow down the sim noticeably, so that's good.

The first few hundred BIOS instructions are starting to match up better with MAME now.

I suspect I could be totally wrong about the MRS/MSR instruction "bug" vs CPU mode, tbh.
And now I'm starting to run your latest version, lots of other things may start working now anyway.

But thanks for taking a look. ;)

**Comment by krevanth** [comment]
@ElectronAsh , 

Happy to help. 

The SV version of the core has the following changes (off the top of my head):
1. Longer pipeline to achieve higher clock speeds.
2. Deeper pipelining of SRAMs to achieve higher clock speeds.
3. Branch Target Buffer to somewhat compensate for increased dependency stalls when increasing pipeline length.

The load latency got worse but there are some enhancements to try and mitigate that (like hit-under-miss cache).

Thanks,
Revanth

**Comment by krevanth** [comment]
I'm very interested in getting a 3DO system working. I'm willing to work together with you to get this thing up. I hope ZAP can (eventually) help you get there.

**Comment by ElectronAsh** [comment]
Thanks.

I've been discussing this mainly on "The 3DO Community" Discord, in the FPGA channel.
No obligation to join, of course, but it may be easier to discuss there.

I just updated the sim_3do repo with your latest SV Zap core.
I think I found one point in the BIOS where it's getting stuck again now.

To me, that suggests it might actually be reading *more* of the correct values, but I need to spoof more stuff to get it past this point again.

If I can even get this latest Zap to boot to the 3DO logo again, I'll be very happy. lol
If it somehow gets past the SWI Overrun error as well, that would be quite big progress.

My code in sim_main.cpp is quite a mess atm. I need to break up into more more functions, but I've often struggled to remember how to use pointers and references etc.

I was just hoping to see something else on the screen other than the logo, and then I could start adding more of the logic to the MADAM and CLIO chips, etc.

I don't think I'll manage to do the CEL (sprite) and polygon rendering stuff by myself, but trapexit, fixel, and others are very knowledgeable on the 3DO hardware. Could probably try translating parts of the Opera 3DO emulator to Verilog/SV.

I just installed Win 11 on a new PC, and setting up MSVC, WSL2, Imgui, and Verilator worked without too many issues.
The main thing really is to point the MSVC project at the correct folders.

And if any of the files that Verilator generates (and puts into the sim_3do/out folder) change, then I just remove them completely from the MSVC project, add them again, then run Debugging.

I haven't tried making MSVC run Verilator for me each time, so I have to just run the ./verilate.sh script from a WSL2 shell.

Anyway, yep, if you have time to hop on the Discord server, we can see what we can do on there. ;)
https://discord.gg/YUfeySVB


ElectronAsh.
